src/goto-symex/symex_assign.cpp

```diff
@@ -658,6 +662,56 @@ void goto_symext::symex_assign_extract(
   symex_assign_rec(ex.from, full_lhs, accuml, full_rhs, guard, hidden);
 }
 
+void goto_symext::symex_assign_bitfield(
+  const expr2tc &lhs,
+  const expr2tc &full_lhs,
+  expr2tc &rhs,
+  expr2tc &full_rhs,
+  guardt &guard,
+  const bool hidden)
+{
+  // Here we expect the LHS to be an expression of the form
+  //   (expr_1 & expr_2) >> expr_3
+  //
+  // Here we basically convert an assignment of the form:
+  //   (val & mask) >> shft := rhs;
+  //
+  // to the form:
+  //   val := new_rhs;
+  //
+  // performing the following 3 steps:
+  //   (1) val := val & not(mask);
+  //   (2) new_val := new_val << shft;
+  //   (3) val := val | new_val;
+  //
+  // where the new LHS is either an index, a byte_extract or
+  // a concat which can be handled recursively
+
+  expr2tc left_side = to_lshr2t(lhs).side_1;
+  expr2tc shft_expr = to_lshr2t(lhs).side_2;
+
+  assert(is_bitand2t(left_side));
+
+  expr2tc value_expr = to_bitand2t(left_side).side_1;
+  expr2tc mask_expr = to_bitand2t(left_side).side_2;
+  expr2tc not_mask_expr = bitnot2tc(mask_expr->type, mask_expr);
+
+  // (1) Resetting the target bits first
+  expr2tc new_value_expr =
+    bitand2tc(value_expr->type, value_expr, not_mask_expr);
+  // (2) Aligning the RHS value with the target bits
+  expr2tc new_rhs = bitcast2tc(shft_expr->type, rhs);
+  new_rhs = shl2tc(new_rhs->type, new_rhs, shft_expr);
+  // (3) Updating the RHS with the remaining bits
+  new_rhs = bitor2tc(new_rhs->type, new_rhs, new_value_expr);
+
+  // Simplify both expression before the recursive call
+  simplify(new_rhs);
+  simplify(value_expr);
```

Comment:
Do you need this here? I think both will be simplified in `symex_assing_symbol`.